cloud-storage/cloud_storage.c

```diff
@@ -0,0 +1,947 @@
+
+static void _upload_file_execute(folder_type_t folder_type, char *file_name);
+
+void _cloud_storage_item_free_fn(void *item)
+{
+   cloud_storage_item_free((cloud_storage_item_t *) item);
+}
+
+void cloud_storage_item_free(cloud_storage_item_t *item)
+{
+   while (item)
+   {
+      free(item->id);
+      free(item->name);
+
+      if (item->item_type == CLOUD_STORAGE_FILE)
+      {
+         if (item->type_data.file.hash_value)
+         {
+            free(item->type_data.file.hash_value);
+         }
+         if (item->type_data.file.download_url)
+         {
+            free(item->type_data.file.download_url);
+         }
+      }
+
+      if (item->item_type == CLOUD_STORAGE_FOLDER)
+      {
+         linked_list_free(item->type_data.folder.children, &_cloud_storage_item_free_fn);
+      }
+
+      free(item);
+   }
+}
+
+static void _free_operation(struct _operation_t *operation)
+{
+   switch (operation->operation_type)
+   {
+      case UPLOAD_FILE:
+         if (operation->parameters.upload.local_file)
+         {
+            free(operation->parameters.upload.local_file);
+         }
+         break;
+      case DOWNLOAD_FILE:
+         if (operation->parameters.download.remote_file)
+         {
+            cloud_storage_item_free(operation->parameters.download.remote_file);
+         }
+         break;
+      default:
+         break;
+   }
+
+   free(operation);
+}
+
+static void _free_operation_fn(void *operation)
+{
+   _free_operation((struct _operation_t *) operation);
+}
+
+static void _operation_thread_loop(void *user_data)
+{
+   while (!_shutdown)
+   {
+      struct _operation_t *operation;
+
+      /* In the critical section:
+       * - Go back to waiting if the opertion queue is empty
+       * - Break out of the loop and shutdown the operation thread if
+       *   _shutdown is true
+       * - Pop an operation off of the queue and release the lock
+       */
+      slock_lock(_operation_queue_mutex);
+      if (generic_queue_length(_operation_queue) == 0)
+      {
+         scond_wait(_operation_condition, _operation_queue_mutex);
+      }
+
+      if (_shutdown)
+      {
+         slock_unlock(_operation_queue_mutex);
+         continue;
+      }
+
+      operation = (struct _operation_t *)generic_queue_pop(_operation_queue);
+      slock_unlock(_operation_queue_mutex);
+
+      if (operation)
+      {
+         switch (operation->operation_type)
+         {
+            case SYNC_FILES:
+               _sync_files_execute(operation->parameters.sync.folder_type);
+               break;
+            case UPLOAD_FILE:
+               _upload_file_execute(operation->parameters.upload.folder_type, operation->parameters.upload.local_file);
+               break;
+            case DOWNLOAD_FILE:
+               break;
+         }
+
+         _free_operation(operation);
+      }
+   }
+
+   /* Broke out of the operation loop, will shutdown the operation thread */
+   if (_operation_queue)
```

Comment:
Wont this always be true? You create the thread after initializing the queue.

Reply:
This was a mistake from changing to the common linked list. Fixed now.